translator/opencensus/oc_to_internal.go

```diff
@@ -0,0 +1,388 @@
+
+		default:
+			str := "<Unknown OpenCensus attribute value type>"
+			values[i].SetString(str)
+		}
+		attrMap[key] = values[i]
+		i++
+	}
+	droppedCount := uint32(ocAttrs.DroppedAttributesCount)
+	return data.NewAttributes(attrMap, droppedCount)
+}
+
+func ocSpanKindToInternal(ocKind octrace.Span_SpanKind, ocAttrs *octrace.Span_Attributes) data.SpanKind {
+	switch ocKind {
+	case octrace.Span_SERVER:
+		return data.SpanKindSERVER
+
+	case octrace.Span_CLIENT:
+		return data.SpanKindCLIENT
+
+	case octrace.Span_SPAN_KIND_UNSPECIFIED:
+		// Span kind field is unspecified, check if TagSpanKind attribute is set.
+		// This can happen if span kind had no equivalent in OC, so we could represent it in
+		// the SpanKind. In that case the span kind may be a special attribute TagSpanKind.
+		if ocAttrs != nil {
+			kindAttr := ocAttrs.AttributeMap[tracetranslator.TagSpanKind]
+			if kindAttr != nil {
+				strVal, ok := kindAttr.Value.(*octrace.AttributeValue_StringValue)
+				if ok && strVal != nil {
+					var otlpKind data.SpanKind
+					switch tracetranslator.OpenTracingSpanKind(truncableStringToStr(strVal.StringValue)) {
+					case tracetranslator.OpenTracingSpanKindConsumer:
+						otlpKind = data.SpanKindCONSUMER
+					case tracetranslator.OpenTracingSpanKindProducer:
+						otlpKind = data.SpanKindPRODUCER
+					default:
+						return data.SpanKindUNSPECIFIED
+					}
+					delete(ocAttrs.AttributeMap, tracetranslator.TagSpanKind)
+					return otlpKind
+				}
+			}
+		}
+		return data.SpanKindUNSPECIFIED
+
+	default:
+		return data.SpanKindUNSPECIFIED
+	}
+}
+
+func ocEventsToInternal(ocEvents *octrace.Span_TimeEvents) (ptrs []*data.SpanEvent, droppedCount uint32) {
+	if ocEvents == nil {
+		return
+	}
+
+	droppedCount = uint32(ocEvents.DroppedMessageEventsCount + ocEvents.DroppedAnnotationsCount)
+
+	eventCount := len(ocEvents.TimeEvent)
+	if eventCount == 0 {
+		return
+	}
+
+	// Create slice that holds pointers and another slice that holds structs at once
+	// to avoid individual allocations of structs.
+	ptrs = make([]*data.SpanEvent, 0, eventCount)
+	content := data.NewSpanEventSlice(eventCount)
+
+	for i, ocEvent := range ocEvents.TimeEvent {
+		if ocEvent == nil {
+			continue
+		}
+
+		// Point one element in slice of pointers to an element in slice of structs.
+		event := &content[i]
+		ptrs = append(ptrs, event)
+
+		event.SetTimestamp(internal.TimestampToUnixnano(ocEvent.Time))
+
+		switch teValue := ocEvent.Value.(type) {
+		case *octrace.Span_TimeEvent_Annotation_:
+			if teValue.Annotation != nil {
+				event.SetName(truncableStringToStr(teValue.Annotation.Description))
+				attrs := ocAttrsToInternal(teValue.Annotation.Attributes)
+				event.SetAttributes(attrs)
+			}
+
+		case *octrace.Span_TimeEvent_MessageEvent_:
+			event.SetAttributes(ocMessageEventToInternalAttrs(teValue.MessageEvent))
+
+		default:
+			event.SetName("An unknown OpenCensus TimeEvent type was detected when translating")
+		}
+	}
+	return
```

Comment:
Shouldn't it be `return ptrs, droppedCount` ?

Reply:
@dmitryax it's equivalent thanks to naked return, see https://tour.golang.org/basics/7

Reply:
Oh right! forgot about that feature